Regression in Liquibase 4.25.0 with H2 DB: Using preconditions "not" + "columnExists" fails with PreconditionErrorException when column not exists

### Search first

- [X] I searched and no similar issues were found

### Description

After Liquibase upgrade from 4.24.0 to 4.25.0, when using it with H2 database, precondition "not" + "columnExists" that worked correctly, now fails with exception `liquibase.precondition.core.ColumnExistsPrecondition@68aec50 : Column "T.MODEL" not found; SQL statement: select t.model from PUBLIC.car t where 0=1` when column not exists.

Precondition that stoped working correctly:
```
<preConditions onFail="MARK_RAN">
      <not><columnExists tableName="car" columnName="model" /></not>
</preConditions>
```

In more details:
With Liquibase 4.25.0 and H2 database, when column not exists, `liquibase.precondition.core.ColumnExistsPrecondition#checkFast` throws `PreconditionErrorException` instead of `PreconditionFailedException`. That is why in `liquibase.precondition.core.NotPrecondition#check` it is not cached and properly handled.

Exception:
```
liquibase.precondition.core.ColumnExistsPrecondition@68aec50 : Column "T.MODEL" not found; SQL statement:
select t.model from PUBLIC.car t where 0=1 [42122-224]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:712) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:119) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:68) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(ChangeLogIterator.java:128) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:184) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:193) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:183) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:162) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:117) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:184) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:193) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:183) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:162) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:250) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.Scope.child(Scope.java:254) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:84) ~[liquibase-core-4.25.0.jar:na]
	... 44 common frames omitted
Caused by: liquibase.exception.PreconditionErrorException: Precondition Error
	at liquibase.precondition.core.ColumnExistsPrecondition.checkFast(ColumnExistsPrecondition.java:183) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.precondition.core.ColumnExistsPrecondition.check(ColumnExistsPrecondition.java:83) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.precondition.core.NotPrecondition.check(NotPrecondition.java:34) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.precondition.core.AndPrecondition.check(AndPrecondition.java:39) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.precondition.core.PreconditionContainer.check(PreconditionContainer.java:213) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:669) ~[liquibase-core-4.25.0.jar:na]
	... 59 common frames omitted
Caused by: liquibase.exception.DatabaseException: org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "T.MODEL" not found; SQL statement:
select t.model from PUBLIC.car t where 0=1 [42122-224]
	at liquibase.database.jvm.JdbcConnection.prepareStatement(JdbcConnection.java:445) ~[liquibase-core-4.25.0.jar:na]
	at liquibase.precondition.core.ColumnExistsPrecondition.checkFast(ColumnExistsPrecondition.java:167) ~[liquibase-core-4.25.0.jar:na]
	... 64 common frames omitted
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "T.MODEL" not found; SQL statement:
select t.model from PUBLIC.car t where 0=1 [42122-224]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:514) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.message.DbException.get(DbException.java:223) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.message.DbException.get(DbException.java:199) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.expression.ExpressionColumn.getColumnException(ExpressionColumn.java:244) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.expression.ExpressionColumn.optimizeOther(ExpressionColumn.java:226) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:213) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.command.query.Select.prepareExpressions(Select.java:1170) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.command.query.Query.prepare(Query.java:218) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.command.Parser.prepareCommand(Parser.java:489) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.engine.SessionLocal.prepareLocal(SessionLocal.java:639) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.engine.SessionLocal.prepareCommand(SessionLocal.java:559) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1166) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:93) ~[h2-2.2.224.jar:2.2.224]
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:316) ~[h2-2.2.224.jar:2.2.224]
	at com.zaxxer.hikari.pool.ProxyConnection.prepareStatement(ProxyConnection.java:337) ~[HikariCP-4.0.3.jar:na]
	at com.zaxxer.hikari.pool.HikariProxyConnection.prepareStatement(HikariProxyConnection.java) ~[HikariCP-4.0.3.jar:na]
	at liquibase.database.jvm.JdbcConnection.prepareStatement(JdbcConnection.java:443) ~[liquibase-core-4.25.0.jar:na]
	... 65 common frames omitted
```

### Steps To Reproduce

1. Create a Spring Boot project with below configuration

pom.xml contains
```
    <dependency>
      <groupId>com.h2database</groupId>
      <artifactId>h2</artifactId>
      <scope>runtime</scope>
    </dependency>
    <dependency>
      <groupId>org.liquibase</groupId>
      <artifactId>liquibase-core</artifactId>
      <version>4.25.0</version>
    </dependency>
```

application.properties
```
spring.liquibase.change-log=classpath:db/changelog/master.xml
```

master.xml
```
<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">

  <changeSet author="XYZ" id="1">
    <createTable tableName="car">
      <column name="id" type="INTEGER">
        <constraints nullable="false" primaryKey="true" primaryKeyName="actor_pkey"/>
      </column>
      <column name="vendor" type="VARCHAR(255)"/>
    </createTable>
  </changeSet>

  <changeSet author="XYZ" id="2">
    <preConditions onFail="MARK_RAN">
      <not><columnExists tableName="car" columnName="model" /></not>
    </preConditions>
    <addColumn tableName="car">
      <column name="model"
        type="VARCHAR(255)"
        defaultValueBoolean="false">
      </column>
    </addColumn>
  </changeSet>

</databaseChangeLog>
```

2. Run `mvn clean package`
3. Migration fails with exception attached earlier

### Expected/Desired Behavior

Full precondition is checked and Liquibase migration succeeds.

### Liquibase Version

4.25.0

### Database Vendor & Version

H2 2.1.214

### Liquibase Integration

maven, spring boot

### Liquibase Extensions

N/A

### OS and/or Infrastructure Type/Provider

N/A

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

--- a/liquibase-integration-tests/src/test/resources/changelogs/common/common.tests.changelog.xml
+++ b/liquibase-integration-tests/src/test/resources/changelogs/common/common.tests.changelog.xml
@@ -240,6 +240,21 @@
         </addColumn>
     </changeSet>
 
+    <changeSet id="column-exists-test" author="januson">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <columnExists columnName="A_NEW_COLUMN" tableName="COLUMN_EXISTS_TEST_TABLE"/>
+            </not>
+        </preConditions>
+        <comment>Testing column exists precondition regression</comment>
+        <createTable tableName="COLUMN_EXISTS_TEST_TABLE">
+            <column name="id" type="int"/>
+        </createTable>
+        <addColumn tableName="COLUMN_EXISTS_TEST_TABLE">
+            <column name="A_NEW_COLUMN" type="int" defaultValueNumeric="1"/>
+        </addColumn>
+    </changeSet>
+
     <changeSet id="tablespace-test1" author="alan">
         <comment>Test tablespace support. Ignored if database does not support tablespaces</comment>
         <createTable tableName="TABLESPACE_TEST_TABLE" tablespace="liquibase2">

--- a/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
+++ b/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
@@ -122,7 +122,6 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
     private void checkFast(Database database, DatabaseChangeLog changeLog)
             throws PreconditionFailedException, PreconditionErrorException {
 
-        PreparedStatement statement = null;
         try {
             String schemaName = getSchemaName();
             if (schemaName == null) {
@@ -133,12 +132,11 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
 
             if (database instanceof PostgresDatabase) {
                 String sql = "SELECT 1 FROM pg_attribute a WHERE EXISTS (SELECT 1 FROM pg_class JOIN pg_catalog.pg_namespace ns ON ns.oid = pg_class.relnamespace WHERE lower(ns.nspname) = ? AND lower(relname) = lower(?) AND pg_class.oid = a.attrelid) AND lower(a.attname) = lower(?);";
-                statement = ((JdbcConnection) database.getConnection())
-                        .prepareStatement(sql);
-                statement.setString(1, schemaName.toLowerCase());
-                statement.setString(2, tableName);
-                statement.setString(3, columnName);
-                try {
+                try (PreparedStatement statement = ((JdbcConnection) database.getConnection())
+                            .prepareStatement(sql)){
+                    statement.setString(1, schemaName.toLowerCase());
+                    statement.setString(2, tableName);
+                    statement.setString(3, columnName);
                     try (ResultSet rs = statement.executeQuery()) {
                         if (rs.next()) {
                             return;
@@ -163,9 +161,8 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
                             database.escapeObjectName(tableName, Table.class));
                 }
 
-                try {
-                    statement = ((JdbcConnection) database.getConnection()).prepareStatement(sql);
-                    statement.executeQuery();
+                try (PreparedStatement statement2 = ((JdbcConnection) database.getConnection()).prepareStatement(sql)){
+                    statement2.executeQuery();
                     // column exists
                 } catch (SQLException | DatabaseException e) {
                     // column or table does not exist
@@ -173,16 +170,10 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
                             "Column %s.%s.%s does not exist", schemaName,
                             tableName, columnName), changeLog, this);
                 }
-                finally {
-                    JdbcUtil.closeStatement(statement);
-                }
             }
 
-        } catch (DatabaseException | SQLException e) {
+        } catch (DatabaseException e) {
             throw new PreconditionErrorException(e, changeLog, this);
-
-        } finally {
-            JdbcUtil.closeStatement(statement);
         }
     }
 

--- a/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
+++ b/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
@@ -131,48 +131,57 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
             String columnName = getColumnName();
 
             if (database instanceof PostgresDatabase) {
-                String sql = "SELECT 1 FROM pg_attribute a WHERE EXISTS (SELECT 1 FROM pg_class JOIN pg_catalog.pg_namespace ns ON ns.oid = pg_class.relnamespace WHERE lower(ns.nspname) = ? AND lower(relname) = lower(?) AND pg_class.oid = a.attrelid) AND lower(a.attname) = lower(?);";
-                try (PreparedStatement statement = ((JdbcConnection) database.getConnection())
-                            .prepareStatement(sql)){
-                    statement.setString(1, schemaName.toLowerCase());
-                    statement.setString(2, tableName);
-                    statement.setString(3, columnName);
-                    try (ResultSet rs = statement.executeQuery()) {
-                        if (rs.next()) {
-                            return;
-                        } else {
-                            // column or table does not exist
-                            throw new PreconditionFailedException(format("Column %s.%s.%s does not exist", schemaName, tableName, columnName), changeLog, this);
-                        }
-                    }
-                } catch (SQLException e) {
-                    throw new PreconditionErrorException(e, changeLog, this);
-                }
+                makeSureColumnExistsInPostgres(database, changeLog, schemaName, tableName, columnName);
             } else {
-                String sql;
-                if (database instanceof FirebirdDatabase) {
-                    sql = format("select t.%s from %s t where 0=1",
-                            database.escapeColumnNameList(columnName),
-                            database.escapeObjectName(tableName, Table.class));
-                } else {
-                    sql = format("select t.%s from %s.%s t where 0=1",
-                            database.escapeColumnNameList(columnName),
-                            database.escapeObjectName(schemaName, Schema.class),
-                            database.escapeObjectName(tableName, Table.class));
-                }
-
-                try (PreparedStatement statement2 = ((JdbcConnection) database.getConnection()).prepareStatement(sql)){
-                    statement2.executeQuery();
-                    // column exists
-                } catch (SQLException | DatabaseException e) {
-                    // column or table does not exist
-                    throw new PreconditionFailedException(format(
-                            "Column %s.%s.%s does not exist", schemaName,
-                            tableName, columnName), changeLog, this);
-                }
+                makeSureColumnExistsInOtherDBs(database, changeLog, schemaName, tableName, columnName);
             }
 
         } catch (DatabaseException e) {
+            throw new PreconditionErrorException(e, changeLog, this);
+        }
+    }
+
+    private void makeSureColumnExistsInOtherDBs(Database database, DatabaseChangeLog changeLog, String schemaName, String tableName, String columnName) throws PreconditionFailedException {
+        String sql;
+        if (database instanceof FirebirdDatabase) {
+            sql = format("select t.%s from %s t where 0=1",
+                    database.escapeColumnNameList(columnName),
+                    database.escapeObjectName(tableName, Table.class));
+        } else {
+            sql = format("select t.%s from %s.%s t where 0=1",
+                    database.escapeColumnNameList(columnName),
+                    database.escapeObjectName(schemaName, Schema.class),
+                    database.escapeObjectName(tableName, Table.class));
+        }
+
+        try (PreparedStatement statement2 = ((JdbcConnection) database.getConnection()).prepareStatement(sql);
+             ResultSet rs = statement2.executeQuery()
+        ){
+            // column exists
+        } catch (SQLException | DatabaseException e) {
+            // column or table does not exist
+            throw new PreconditionFailedException(format(
+                    "Column %s.%s.%s does not exist", schemaName,
+                    tableName, columnName), changeLog, this);
+        }
+    }
+
+    private void makeSureColumnExistsInPostgres(Database database, DatabaseChangeLog changeLog, String schemaName, String tableName, String columnName) throws PreconditionFailedException, PreconditionErrorException, DatabaseException {
+        String sql = "SELECT 1 FROM pg_attribute a WHERE EXISTS (SELECT 1 FROM pg_class JOIN pg_catalog.pg_namespace ns ON ns.oid = pg_class.relnamespace WHERE lower(ns.nspname) = ? AND lower(relname) = lower(?) AND pg_class.oid = a.attrelid) AND lower(a.attname) = lower(?);";
+        try (PreparedStatement statement = ((JdbcConnection) database.getConnection())
+                    .prepareStatement(sql)){
+            statement.setString(1, schemaName.toLowerCase());
+            statement.setString(2, tableName);
+            statement.setString(3, columnName);
+            try (ResultSet rs = statement.executeQuery()) {
+                if (rs.next()) {
+                    return;
+                } else {
+                    // column or table does not exist
+                    throw new PreconditionFailedException(format("Column %s.%s.%s does not exist", schemaName, tableName, columnName), changeLog, this);
+                }
+            }
+        } catch (SQLException e) {
             throw new PreconditionErrorException(e, changeLog, this);
         }
     }

--- a/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
+++ b/liquibase-standard/src/main/java/liquibase/precondition/core/ColumnExistsPrecondition.java
@@ -163,12 +163,11 @@ public class ColumnExistsPrecondition extends AbstractPrecondition {
                             database.escapeObjectName(tableName, Table.class));
                 }
 
-                statement = ((JdbcConnection) database.getConnection())
-                        .prepareStatement(sql);
                 try {
-                    statement.executeQuery().close();
+                    statement = ((JdbcConnection) database.getConnection()).prepareStatement(sql);
+                    statement.executeQuery();
                     // column exists
-                } catch (SQLException e) {
+                } catch (SQLException | DatabaseException e) {
                     // column or table does not exist
                     throw new PreconditionFailedException(format(
                             "Column %s.%s.%s does not exist", schemaName,

